clang/runtime/dpct-rt/include/dpl_extras/dpcpp_extensions.h

```diff
@@ -399,6 +400,34 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
     }
   }
 
+  /// Rearrange elements from blocked order to striped order
+  template <typename Item>
+  __dpct_inline__ void blocked_to_striped(Item item,
+                                          T (&keys)[VALUES_PER_THREAD],
+                                          int (&ranks)[VALUES_PER_THREAD]) {
+    T *buffer = reinterpret_cast<T *>(_local_memory);
+
+#pragma unroll
+    for (int i = 0; i < VALUES_PER_THREAD; i++) {
+      int offset = (item.get_local_id(0) * VALUES_PER_THREAD) + i;
```

Comment:
Can we consolidate redundant code between `blocked_to_striped` and `scatter_to_blocked`?
The only place they differ is in generating the offsets.  
Either by passing lambdas or having constexpr branches and some named offset generating helpers,  I think this could be consolidated.

Reply:
@danhoeflinger  yes I have refactored the code to better align with this design. Could you please take a look . Thanks

Reply:
I don't really see this as resolved, but perhaps it is more trouble than it is worth.  Let me better explain my intentions, and you can decide if its helpful or if its adding too much complication in order to remove the redundancy.  I don't see it as required.

Right now we have 3 copies of effectively the same function, the difference between them is the calculation of the incoming load offset index and the calculation of the outgoing store offset index.  
Moreover, for blocked to striped and striped to blocked, the offset index calculation for the input matches the offset index calculation of the output for the other and vice versa.  

Effectively you have one function helper (pseudocode):
```
helper_exchange(normal_inputs, offset_functor_input, offset_functor_output)
{
    reinterpret exchange temporary memory
    unrolled normal load loop, but calling offset_functor_input to get input index from item and `i`
    
    barrier

    unrolled normal store loop but calling offset_functor_output to get output index from item
}
```
Then you would have a struct functor definition for the blocked index calculation, and another for the striped index calculation from `item` and `i`.  

To reuse the building block for the scatter to blocked, you'd also need to create a functor to obtain the offset from `ranks`.

Finally, you would have the external APIs which use those building blocks to implement the general APIs.